protocol/vm/introspection.go

```diff
@@ -50,207 +38,163 @@ func opCheckOutput(vm *virtualMachine) error {
+}
 
+func opProgram(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	return vm.push(vm.mainprog, true)
+	return vm.push(vm.context.Code, true)
 }
 
 func opMinTime(vm *virtualMachine) error {
-	if vm.tx == nil {
-		return ErrContext
-	}
-
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	return vm.pushInt64(int64(vm.tx.MinTime), true)
-}
-
-func opMaxTime(vm *virtualMachine) error {
-	if vm.tx == nil {
+	if vm.context.MinTimeMS == nil {
 		return ErrContext
 	}
+	return vm.pushInt64(int64(*vm.context.MinTimeMS), true)
+}
 
+func opMaxTime(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	maxTime := vm.tx.MaxTime
-	if maxTime == 0 || maxTime > math.MaxInt64 {
-		maxTime = uint64(math.MaxInt64)
+	if vm.context.MaxTimeMS == nil {
+		return ErrContext
+	}
+	maxTimeMS := *vm.context.MaxTimeMS
+	if maxTimeMS == 0 || maxTimeMS > math.MaxInt64 {
+		maxTimeMS = uint64(math.MaxInt64)
 	}
 
-	return vm.pushInt64(int64(maxTime), true)
+	return vm.pushInt64(int64(maxTimeMS), true)
 }
 
 func opRefDataHash(vm *virtualMachine) error {
-	if vm.tx == nil {
-		return ErrContext
-	}
-
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	h := sha3.Sum256(vm.tx.Inputs[vm.inputIndex].ReferenceData)
-	return vm.push(h[:], true)
-}
-
-func opTxRefDataHash(vm *virtualMachine) error {
-	if vm.tx == nil {
+	if vm.context.InputRefDataHash == nil {
 		return ErrContext
 	}
+	return vm.push(*vm.context.InputRefDataHash, true)
+}
 
+func opTxRefDataHash(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	h := sha3.Sum256(vm.tx.ReferenceData)
-	return vm.push(h[:], true)
-}
-
-func opIndex(vm *virtualMachine) error {
-	if vm.tx == nil {
+	if vm.context.TxRefDataHash == nil {
 		return ErrContext
 	}
+	return vm.push(*vm.context.TxRefDataHash, true)
+}
 
+func opIndex(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	return vm.pushInt64(int64(vm.inputIndex), true)
-}
-
-func opOutputID(vm *virtualMachine) error {
-	if vm.tx == nil {
-		return ErrContext
-	}
-
-	outid := vm.txContext.OutputID
-	if outid == nil {
+	if vm.context.InputIndex == nil {
 		return ErrContext
 	}
+	return vm.pushInt64(int64(*vm.context.InputIndex), true)
+}
 
+func opOutputID(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	return vm.push(outid[:], true)
-}
-
-func opNonce(vm *virtualMachine) error {
-	if vm.tx == nil {
-		return ErrContext
-	}
-
-	txin := vm.tx.Inputs[vm.inputIndex]
-	ii, ok := txin.TypedInput.(*bc.IssuanceInput)
-	if !ok {
+	if vm.context.SpentOutputID == nil {
 		return ErrContext
 	}
+	return vm.push(*vm.context.SpentOutputID, true)
+}
 
+func opNonce(vm *virtualMachine) error {
 	err := vm.applyCost(1)
 	if err != nil {
 		return err
 	}
 
-	return vm.push(ii.Nonce, true)
+	if vm.context.Nonce == nil {
```

Comment:
It's a pretty inconsequential behavior change that only manifests when both error conditions are true (it's the wrong context _and_ the runlimit is exhausted). The old code would detect one of the errors first, the new code detects the other. (It still works the same way in the case of zero errors or one.) I don't think it's worth mentioning or changing.